### PR TITLE
Error arena optimization

### PR DESCRIPTION
**Pull request by ncteisen**
Built on #9756.

This optimization removes the avl trees from grpc_error. The ints, strs, times, and errs are now stored in one arene, with byte arrays indexing into it if they are present. 

This optimization also holds the strings as gprc_slice's. For now, all are copied and reallocated. But in the next optimization, I will pull that up a level and use grpc_slice_from_static_string where appropriate.

On Linux opt build, error create went from 587 nanoseconds to 148. On Mac it dropped from 1671 to 231. Notably, the allocations per new error for both Linux and Mac dropped from 10 to 1. And the atomic adds when from 66 to 7!

### CODE REVIEW

**Comment by ncteisen** [comment]
I *think* that this PR and its predecessor, #9756, are ready for review. All of the red looks unrelated.

File: test/core/iomgr/error_test.c [hunk 1]
```diff
@@ -0,0 +1,219 @@
+/*
+ *
+ * Copyright 2015, Google Inc.
```
Comment (ctiller) [inline test/core/iomgr/error_test.c]: Wait, what year is it?

File: src/core/lib/iomgr/error_internal.h [hunk 2]
```diff
@@ -35,18 +35,28 @@
 #define GRPC_CORE_LIB_IOMGR_ERROR_INTERNAL_H
 
 #include <inttypes.h>
-#include <stdbool.h>
+#include <stdbool.h>  // TODO, do we need this?
 
-#include <grpc/support/avl.h>
+#include <grpc/support/sync.h>
+
+typedef struct linked_error linked_error;
```
Comment (ctiller) [inline src/core/lib/iomgr/error_internal.h]: grpc_linked_error (we need to prefix all names in the global namespace)

File: test/core/iomgr/error_test.c (hunk 1 above)
Comment (ncteisen) [inline test/core/iomgr/error_test.c]: Oops, done.

File: src/core/lib/iomgr/error_internal.h (hunk 2 above)
Comment (ncteisen) [inline src/core/lib/iomgr/error_internal.h]: Done.

**Comment by ncteisen** [comment]
Interop Failures: #9937, #9878, #10056
Mac Failure: #9550
Linux Failure: #9185

**Pull request merged**